File: language/dsl/fluent/src/core/base-builder/storage/auxiliary-storage.ts
/**
 * Manages auxiliary data storage for builders
 *
 * Auxiliary data is metadata that doesn't appear in the final built object
 * but is used during the build process. This includes:
 * - Templates (stored under "__templates__")
 * - Switches (stored under "__switches__")
 * - Custom metadata
 *
 * This separation keeps builder state clean and makes the build process more explicit
 */
export class AuxiliaryStorage {
  private data: Map<string, unknown> = new Map();

  /**
   * Sets auxiliary data with a given key
   */
  set<T>(key: string, value: T): void {
    this.data.set(key, value);
  }

  /**
   * Gets auxiliary data with type assertion
   * The caller is responsible for knowing the correct type
   */
  get<T>(key: string): T | undefined {
    const value = this.data.get(key);
    return value as T | undefined;
  }

  /**
   * Pushes an item to an auxiliary data array
   * Creates the array if it doesn't exist
   */
  push<T>(key: string, item: T): void {
    const existing = this.data.get(key);
    if (Array.isArray(existing)) {
      existing.push(item);
    } else {
      this.data.set(key, [item]);
    }
  }

  /**
   * Gets an auxiliary data array
   * Returns empty array if doesn't exist or isn't an array
   */
  getArray<T>(key: string): T[] {
    const existing = this.data.get(key);
    return Array.isArray(existing) ? existing : [];
  }

  /**
   * Checks if auxiliary data exists for a key
   */
  has(key: string): boolean {
    return this.data.has(key);
  }

  /**
   * Deletes auxiliary data for a key
   */
  delete(key: string): boolean {
    return this.data.delete(key);
  }

  /**
   * Clears all auxiliary data
   */
  clear(): void {
    this.data.clear();
  }

  /**
   * Clones the auxiliary storage, creating an independent copy
   */
  clone(): AuxiliaryStorage {
    const cloned = new AuxiliaryStorage();
    cloned.data = new Map(this.data);
    return cloned;
  }
}

File: language/dsl/kotlin/src/main/kotlin/com/intuit/playertools/fluent/core/AuxiliaryStorage.kt
package com.intuit.playertools.fluent.core

/**
 * A type-safe key for AuxiliaryStorage.
 * The type parameter T represents the type of value stored at this key.
 */
class TypedKey<T : Any>(
    val name: String,
)

/**
 * A type-safe key for list values in AuxiliaryStorage.
 * The type parameter T represents the element type of the list.
 */
class TypedListKey<T : Any>(
    val name: String,
)

/**
 * Storage for auxiliary metadata like templates and switches.
 * Uses typed keys to ensure type safety at call sites.
 */
class AuxiliaryStorage {
    private val data = mutableMapOf<String, Any>()

    /**
     * Sets a value for a typed key, replacing any existing value.
     */
    operator fun <T : Any> set(
        key: TypedKey<T>,
        value: T,
    ) {
        data[key.name] = value
    }

    /**
     * Gets a value by typed key.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : Any> get(key: TypedKey<T>): T? = data[key.name] as? T

    /**
     * Pushes an item to a list at the given typed list key.
     * Creates a new list if one doesn't exist.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : Any> push(
        key: TypedListKey<T>,
        item: T,
    ) {
        val existing = data[key.name]
        if (existing is MutableList<*>) {
            (existing as MutableList<T>).add(item)
        } else {
            data[key.name] = mutableListOf(item)
        }
    }

    /**
     * Gets a list by typed list key, returning empty list if not found.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : Any> getList(key: TypedListKey<T>): List<T> = (data[key.name] as? List<T>) ?: emptyList()

    /**
     * Checks if a typed key exists.
     */
    fun <T : Any> has(key: TypedKey<T>): Boolean = key.name in data

    /**
     * Checks if a typed list key exists.
     */
    fun <T : Any> has(key: TypedListKey<T>): Boolean = key.name in data

    /**
     * Removes a value by typed key.
     */
    fun <T : Any> remove(key: TypedKey<T>): Boolean = data.remove(key.name) != null

    /**
     * Removes a value by typed list key.
     */
    fun <T : Any> remove(key: TypedListKey<T>): Boolean = data.remove(key.name) != null

    /**
     * Clears all stored data.
     */
    fun clear() = data.clear()

    /**
     * Creates a shallow clone of this storage.
     */
    fun clone(): AuxiliaryStorage =
        AuxiliaryStorage().also { cloned ->
            cloned.copyFrom(this)
        }

    /**
     * Copies all data from another AuxiliaryStorage instance.
     * Clears existing data before copying.
     */
    fun copyFrom(other: AuxiliaryStorage) {
        data.clear()
        other.data.forEach { (key, value) ->
            data[key] =
                if (value is MutableList<*>) {
                    value.toMutableList()
                } else {
                    value
                }
        }
    }

    companion object {
        /** Key for storing template configuration functions */
        internal val TEMPLATES = TypedListKey<(BuildContext) -> TemplateConfig>("__templates__")

        /** Key for storing switch metadata */
        internal val SWITCHES = TypedListKey<SwitchMetadata>("__switches__")
    }
}
